File: apps/api/src/outreachos_bridge.py
"""JSON stdin/stdout adapter for the local OutreachOS Python source tree.

The TypeScript API owns HTTP; this runner imports the Python engine directly from
OUTREACHOS_ROOT/src and returns compact, JSON-safe results.  It does not start an
HTTP server or depend on the separate OutreachOS dashboard process.
"""
from __future__ import annotations

import json
import os
import sys
import traceback
from pathlib import Path
from typing import Any


def _fail(code: str, message: str, detail: str | None = None) -> dict[str, Any]:
    result: dict[str, Any] = {"ok": False, "error": {"code": code, "message": message}}
    if detail:
        result["error"]["detail"] = detail
    return result


def _engine():
    root = os.environ.get("OUTREACHOS_ROOT", "").strip()
    if not root:
        raise RuntimeError("OUTREACHOS_ROOT is required")
    source = Path(root).expanduser().resolve() / "src"
    if not source.is_dir():
        raise RuntimeError(f"OutreachOS source missing: {source}")
    sys.path.insert(0, str(source))
    from outreachos.config import SETTINGS
    from outreachos.orchestration.engine import Engine
    from outreachos.pool.store import PoolStore

    db_path = os.environ.get("OUTREACHOS_DB", "").strip() or str(Path(root).expanduser().resolve() / "outreachos.db")
    return Engine(PoolStore(db_path)), SETTINGS, db_path


def _campaigns(engine) -> list[dict[str, Any]]:
    return [
        {
            "id": campaign.id,
            "name": campaign.name,
            "status": campaign.status,
            "clientId": campaign.client_id,
            "icp": campaign.icp.to_dict(),
            "createdAt": campaign.created_at,
        }
        for campaign in engine.active_campaigns()
    ]


def _search_leads(engine, campaign: str, limit: int) -> dict[str, Any]:
    selected = engine.get_campaign(campaign)
    leads = engine.store.leads(selected.id)
    return {
        "campaign": selected.name,
        "count": len(leads),
        "leads": [
            {
                "id": lead.id,
                "name": lead.full_name,
                "title": lead.title,
                "company": lead.company,
                "email": lead.email,
                "emailStatus": lead.email_status,
                "stage": lead.stage,
                "outreachState": lead.outreach_state,
                "location": lead.location,
                "industry": lead.industry,
                "intentScore": lead.enrichment.get("intent_score", 0),
                "triggerSignal": lead.trigger_signal,
            }
            for lead in leads[:limit]
        ],
    }


def _operation(payload: dict[str, Any]) -> dict[str, Any]:
    operation = payload.get("operation")
    engine, settings, db_path = _engine()
    if operation == "status":
        return {
            "ok": True,
            "data": {
                "available": True,
                "source": "outreachos-python-engine",
                "dbPath": db_path,
                "providerMode": settings.provider_mode,
                "llmMode": settings.llm_mode,
                "writeEnabled": os.environ.get("OUTREACHOS_BRIDGE_WRITE_ENABLED", "false").lower() == "true",
            },
        }
    if operation == "overview":
        campaigns = []
        totals = {"leads": 0, "booked": 0, "sent": 0}
        for campaign in engine.active_campaigns():
            stats = engine.stats(campaign.name)
            booked = stats["by_outreach_state"].get("booked", 0)
            totals["leads"] += stats["total_leads"]
            totals["booked"] += booked
            totals["sent"] += stats["sent_est"]
            campaigns.append({"id": campaign.id, "name": campaign.name, "status": campaign.status, "stats": stats})
        return {"ok": True, "data": {"totals": totals, "campaigns": campaigns}}
    if operation == "campaigns":
        return {"ok": True, "data": {"campaigns": _campaigns(engine)}}
    campaign = str(payload.get("campaign", "")).strip()
    if operation == "campaign_stats":
        return {"ok": True, "data": engine.stats(campaign)}
    if operation == "search_leads":
        return {"ok": True, "data": _search_leads(engine, campaign, min(max(int(payload.get("limit", 50)), 1), 100))}
    if operation == "run_cycle":
        if os.environ.get("OUTREACHOS_BRIDGE_WRITE_ENABLED", "false").lower() != "true":
            return _fail("BRIDGE_WRITE_DISABLED", "Cycle execution is disabled for this bridge runtime")
        return {"ok": True, "data": engine.full_cycle(campaign, limit=min(max(int(payload.get("limit", 25)), 1), 100))}
    return _fail("UNKNOWN_OPERATION", f"Unknown bridge operation: {operation}")


def main() -> None:
    try:
        payload = json.load(sys.stdin)
        if not isinstance(payload, dict):
            raise ValueError("Request must be a JSON object")
        result = _operation(payload)
    except KeyError as error:
        result = _fail("NOT_FOUND", str(error))
    except Exception as error:  # structured error lets the TypeScript API keep its HTTP contract
        result = _fail("OUTREACHOS_BRIDGE_ERROR", str(error), traceback.format_exc(limit=3))
    print(json.dumps(result, default=str))


if __name__ == "__main__":
    main()

File: apps/api/src/outreachos-bridge.ts
import { existsSync } from "node:fs";
import { basename, dirname, resolve } from "node:path";
import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";

export type OutreachOSOperation = "status" | "overview" | "campaigns" | "campaign_stats" | "search_leads" | "run_cycle" | "run_agent";
export type OutreachOSBridgeStatus = {
  configured: boolean;
  root?: string;
  python?: string;
  writeEnabled: boolean;
  reason?: string;
};

export type OutreachOSRuntimeBridge = {
  status(): OutreachOSBridgeStatus;
  call(operation: OutreachOSOperation, input?: Record<string, unknown>): Promise<Record<string, unknown>>;
};

type BridgeOptions = {
  root?: string;
  python?: string;
  writeEnabled?: boolean;
  timeoutMs?: number;
  runnerPath?: string;
};

const moduleDirectory = dirname(fileURLToPath(import.meta.url));
const localRootCandidates = [
  resolve(process.cwd(), "../../OPENCODE_PROJECTS/OutreachOS"),
  resolve(process.cwd(), "../../../OPENCODE_PROJECTS/OutreachOS"),
  resolve(process.cwd(), "../../../../OPENCODE_PROJECTS/OutreachOS"),
  resolve(moduleDirectory, "../../../../../OPENCODE_PROJECTS/OutreachOS"),
];

const discoverLocalRoot = (): string | undefined => localRootCandidates.find((candidate) => existsSync(resolve(candidate, "src")));

export class PythonOutreachOSBridge implements OutreachOSRuntimeBridge {
  private readonly root: string | undefined;
  private readonly python: string;
  private readonly writeEnabled: boolean;
  private readonly timeoutMs: number;
  private readonly runnerPath: string;

  constructor(options: BridgeOptions = {}) {
    const configuredRoot = options.root ?? process.env.OUTREACHOS_ROOT;
    const root = configuredRoot || discoverLocalRoot();
    this.root = root && existsSync(resolve(root, "src")) ? resolve(root) : undefined;
    this.python = options.python ?? process.env.OUTREACHOS_PYTHON ?? (this.root && existsSync(resolve(this.root, ".venv/bin/python")) ? resolve(this.root, ".venv/bin/python") : "python3");
    this.writeEnabled = options.writeEnabled ?? process.env.OUTREACHOS_BRIDGE_WRITE_ENABLED === "true";
    this.timeoutMs = options.timeoutMs ?? 20_000;
    this.runnerPath = options.runnerPath ?? resolve(moduleDirectory, "outreachos_bridge.py");
  }

  status(): OutreachOSBridgeStatus {
    if (!this.root) return { configured: false, writeEnabled: false, reason: "Set OUTREACHOS_ROOT to an OutreachOS source checkout" };
    if (!existsSync(this.runnerPath)) return { configured: false, writeEnabled: false, reason: `Bridge runner missing: ${basename(this.runnerPath)}` };
    return { configured: true, root: this.root, python: this.python, writeEnabled: this.writeEnabled };
  }

  async call(operation: OutreachOSOperation, input: Record<string, unknown> = {}): Promise<Record<string, unknown>> {
    const state = this.status();
    if (!state.configured || !this.root) throw new OutreachOSBridgeError("OUTREACHOS_NOT_CONFIGURED", state.reason ?? "OutreachOS bridge is unavailable", 503);
    if (operation === "run_cycle" && !this.writeEnabled) throw new OutreachOSBridgeError("BRIDGE_WRITE_DISABLED", "Cycle execution is disabled for this bridge runtime", 403);
    const response = await this.execute({ operation, ...input });
    if (response.ok !== true) {
      const error = asRecord(response.error);
      throw new OutreachOSBridgeError(stringValue(error.code, "OUTREACHOS_BRIDGE_ERROR"), stringValue(error.message, "OutreachOS bridge request failed"), error.code === "NOT_FOUND" ? 404 : 502);
    }
    return asRecord(response.data);
  }

  private async execute(payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    return new Promise((resolvePromise, rejectPromise) => {
      const child = spawn(this.python!, [this.runnerPath], {
        env: {
          ...process.env,
          OUTREACHOS_ROOT: this.root,
          OUTREACHOS_DB: process.env.OUTREACHOS_DB ?? resolve(this.root!, "outreachos.db"),
          OUTREACHOS_BRIDGE_WRITE_ENABLED: String(this.writeEnabled),
        },
        stdio: ["pipe", "pipe", "pipe"],
      });
      let stdout = "";
      let stderr = "";
      const timeout = setTimeout(() => child.kill("SIGTERM"), this.timeoutMs);
      child.stdout.on("data", (chunk: Buffer) => { stdout += chunk.toString(); });
      child.stderr.on("data", (chunk: Buffer) => { stderr += chunk.toString(); });
      child.once("error", (error) => { clearTimeout(timeout); rejectPromise(new OutreachOSBridgeError("OUTREACHOS_PROCESS_ERROR", error.message, 502)); });
      child.once("close", (code) => {
        clearTimeout(timeout);
        if (code !== 0) {
          rejectPromise(new OutreachOSBridgeError("OUTREACHOS_PROCESS_ERROR", stderr.trim() || `OutreachOS runner exited ${code}`, 502));
          return;
        }
        try {
          resolvePromise(asRecord(JSON.parse(stdout)));
        } catch {
          rejectPromise(new OutreachOSBridgeError("OUTREACHOS_PROCESS_ERROR", `OutreachOS runner returned invalid JSON${stderr ? `: ${stderr.trim()}` : ""}`, 502));
        }
      });
      child.stdin.end(JSON.stringify(payload));
    });
  }
}

export class OutreachOSBridgeError extends Error {
  constructor(public readonly code: string, message: string, public readonly statusCode: number) {
    super(message);
    this.name = "OutreachOSBridgeError";
  }
}

const asRecord = (value: unknown): Record<string, unknown> => value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {};
const stringValue = (value: unknown, fallback: string): string => typeof value === "string" && value ? value : fallback;
